components/ILIAS/Certificate/classes/Setup/Migration/CertificateIdMigration.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
+ * published by ILIAS open source e-Learning e.V.
+ *
+ * ILIAS is licensed with the GPL-3.0,
+ * see https://www.gnu.org/licenses/gpl-3.0.en.html
+ * You should have received a copy of said license along with the
+ * source code, too.
+ *
+ * If this is not the case or you just want to try ILIAS, you'll find
+ * us at:
+ * https://www.ilias.de
+ * https://github.com/ILIAS-eLearning
+ *
+ *********************************************************************/
+
+declare(strict_types=1);
+
+
+namespace ILIAS\Certificate\Setup\Migration;
+
+use ilDatabaseException;
+use ilDatabaseUpdatedObjective;
+use ilDBConstants;
+use ilDBPdoInterface;
+use ilDBStatement;
+use ILIAS\Setup\Environment;
+use ILIAS\Setup\Migration;
+use JsonException;
+use ReflectionClass;
+use ILIAS\Data\UUID\Factory;
+
+
+class CertificateIdMigration implements Migration
+{
+    public const NUMBER_OF_STEPS = 10000;
+
+    private ilDBPdoInterface $db;
+    private Factory $uuid_factory;
+    private ilDBStatement $prepared_statement;
+
+    public function getLabel(): string
+    {
+        return (new ReflectionClass($this))->getShortName();
+    }
+    public function getDefaultAmountOfStepsPerRun(): int
+    {
+        return self::NUMBER_OF_STEPS;
+    }
+
+    public function getPreconditions(Environment $environment): array
+    {
+        return [
+            new ilDatabaseUpdatedObjective()
+        ];
+    }
+
+    public function prepare(Environment $environment): void
+    {
+        $this->db = $environment->getResource(Environment::RESOURCE_DATABASE);
+        $this->prepared_statement = $this->db->prepareManip(
+            'UPDATE il_cert_user_cert SET certificate_id = ?, template_values = ? WHERE id = ?',
+            [
+                ilDBConstants::T_TEXT,
+                ilDBConstants::T_TEXT,
+                ilDBConstants::T_INTEGER
+            ]
+        );
+
+        $this->uuid_factory = new Factory();
+    }
+
+    /**
+     * @throws JsonException
+     * @throws ilDatabaseException
+     */
+    public function step(Environment $environment): void
+    {
+        $result = $this->db->query(
```

Comment:
Please have a look at the consumer code of set interface methods in `\ILIAS\Setup\Objective\MigrationObjective::achieve`.

* `getRemainingAmountOfSteps` defines the number of iterations/steps of a `while` loop`.
* `step` should run a single step determined by `getRemainingAmountOfSteps`

So potentially you tell the consumer, there are `1_000_000` steps (certificates without an ID), but you update all of them in one single step.

So IMO we should interpret the steps as some kind of "page" (a list of items) a "page" can contain multiple records we'll have to update.